### PR TITLE
[SDK-228] BaseService Interface added.

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

File: Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs [hunk 1]
```diff
@@ -1,29 +1,46 @@
...
             _countlyUtils = countlyUtils;
+            _eventCountlyService = eventCountlyService;
+            _requestCountlyHelper = requestCountlyHelper;
+            _sessionCountlyService = sessionCountlyService;
+            _listeners = new List<IBaseService>();
         }
 
         public string DeviceId { get; private set; }
 
+        internal void AddLitener(IBaseService listener)
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs:31]: typo

File: Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs [hunk 2]
```diff
@@ -32,12 +49,19 @@ internal void InitDeviceId(string deviceId = null)
...
-                ? storedDeviceId
-                : !_countlyUtils.IsNullEmptyOrWhitespace(DeviceId)
-                    ? DeviceId
-                    : !_countlyUtils.IsNullEmptyOrWhitespace(deviceId)
-                        ? deviceId : _countlyUtils.GetUniqueDeviceId();
+            if (!_countlyUtils.IsNullEmptyOrWhitespace(storedDeviceId)) {
+                DeviceId = storedDeviceId;
+            } else {
+                if (!_countlyUtils.IsNullEmptyOrWhitespace(DeviceId)) {
+                    return;
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs:56]: why is this "return" here? There was none in the original implementation.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
The current implementation does not represent the requested thing from our communication. 

Countly.cs should have a list of IBaseService objects as a class variable.
That list should be populated at the end of init.
That whole list should then be provided to the device ID service.

File: Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs (hunk 2 above)
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs:56]: do i need to add this statement?
DeviceId = DeviceId

Comment (ArtursKadikis) [reply Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs:56]: No, but do it without the "return" statement

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Changes not implemented

File: Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs (hunk 2 above)
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs]: I have removed the return statement.

File: Assets/Plugins/CountlySDK/Countly.cs [hunk 3]
```diff
@@ -216,18 +218,35 @@ public async void Init(CountlyConfiguration configuration)
...
+            Device = new DeviceIdCountlyService(Configuration, Session, requests, Events, countlyUtils);
             _inputObserver = InputObserverResolver.Resolve();
+
+            RegisterServicesToDeviceService();
         }
 
+        private void RegisterServicesToDeviceService() {
+            _listeners.Clear();
+
+            _listeners.Add(Consents);
```
Comment (ArtursKadikis) [inline Assets/Plugins/CountlySDK/Countly.cs:236]: All of these listeners should be added to the array independently of the device ID service.
Either rename this function to something more generic or move all of the "_listeners.Add(...)" outside of this function.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

File: Assets/Plugins/CountlySDK/Countly.cs (hunk 3 above)
Comment (ZahidZafar) [reply Assets/Plugins/CountlySDK/Countly.cs:236]: "All of these listeners should be added to the array independently of the device ID service."

We have to define the size of the array at the start,
independently of the device ID service --> need more words to understand.
Do you mean DeviceId Service tell countly.cs class on device id change and countly.cs will tell all other listeners(services)?

**Review by ArtursKadikis — APPROVED** [review]

**Pull request merged**